File: src/packet/payload/car_telemetry.rs
use bytemuck::{Pod, Zeroable};

use crate::packet::header;

#[repr(C, packed)]
#[derive(Clone, Copy, Debug, Zeroable, Pod)]
pub struct CarTelemetry {
    pub m_speed: u16,
    pub m_throttle: f32,
    pub m_steer: f32,
    pub m_brake: f32,
    pub m_clutch: u8,
    pub m_gear: i8,
    pub m_engine_rpm: u16,
    pub m_drs: u8,
    pub m_rev_lights_percent: u8,
    pub m_rev_lights_bit_value: u16,
    pub m_brakes_temperature: [u16; 4],
    pub m_tyres_surface_temperature: [u8; 4],
    pub m_tyres_inner_temperature: [u8; 4],
    pub m_engine_temperature: u16,
    pub m_tyres_pressure: [f32; 4],
    pub m_surface_type: [u8; 4],
}

#[repr(C, packed)]
#[derive(Clone, Copy, Debug, Zeroable, Pod)]
pub struct PacketCarTelemetry {
    pub m_header: header::PacketHeader,
    pub m_car_telemetry_data: [CarTelemetry; 22],
    pub m_mfd_panel_index: u8,
    pub m_mfd_panel_index_secondary_player: u8,
    pub m_suggested_gear: i8,
}

File: consumer/src/main/scala/packet/payload/CarTelemetry.scala
package packet.payload

import packet.Header

case class CarTelemetry(
    m_speed: Int,
    m_throttle: Float,
    m_steer: Float,
    m_brake: Float,
    m_clutch: Int,
    m_gear: Int,
    m_engine_rpm: Int,
    m_drs: Int,
    m_rev_lights_percent: Int,
    m_rev_lights_bit_value: Int,
    m_brakes_temperature: Array[Int],
    m_tyres_surface_temperature: Array[Int],
    m_tyres_inner_temperature: Array[Int],
    m_engine_temperature: Int,
    m_tyres_pressure: Array[Float],
    m_surface_type: Array[Int]
)

case class PacketCarTelemetry(
    m_header: Header,
    m_car_telemetry_data: Array[CarTelemetry],
    m_mfd_panel_index: Int,
    m_mfd_panel_index_secondary_player: Int,
    m_suggested_gear: Int
)
